module/zfs/qat.h

```diff
@@ -0,0 +1,191 @@
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+#ifndef	_SYS_QAT_H
+#define	_SYS_QAT_H
+
+typedef enum qat_compress_dir {
+	QAT_DECOMPRESS = 0,
+	QAT_COMPRESS = 1,
+} qat_compress_dir_t;
+
+typedef enum qat_encrypt_dir {
+	QAT_DECRYPT = 0,
+	QAT_ENCRYPT = 1,
+} qat_encrypt_dir_t;
+
+
+#if defined(_KERNEL) && defined(HAVE_QAT)
+#include <sys/zio.h>
+#include <sys/crypto/api.h>
+#include "cpa.h"
+#include "dc/cpa_dc.h"
+#include "lac/cpa_cy_sym.h"
+
+/*
+ * Timeout - no response from hardware after 0.5 seconds
+ */
+#define	QAT_TIMEOUT_MS		500
+
+/*
+ * The minimal and maximal buffer size, which are not restricted
+ * in the QAT hardware, but with the input buffer size between 4KB
+ * and 128KB, the hardware can provide the optimal performance.
+ */
+#define	QAT_MIN_BUF_SIZE	(4*1024)
+#define	QAT_MAX_BUF_SIZE	(128*1024)
+
+/*
+ * Used for qat kstat.
+ */
+typedef struct qat_stats {
+	/*
+	 * Number of jobs submitted to qat compression engine.
+	 */
+	kstat_named_t comp_requests;
+	/*
+	 * Total bytes sent to qat compression engine.
+	 */
+	kstat_named_t comp_total_in_bytes;
+	/*
+	 * Total bytes output from qat compression engine.
+	 */
+	kstat_named_t comp_total_out_bytes;
+	/*
+	 * Number of jobs submitted to qat de-compression engine.
+	 */
+	kstat_named_t decomp_requests;
+	/*
+	 * Total bytes sent to qat de-compression engine.
+	 */
+	kstat_named_t decomp_total_in_bytes;
+	/*
+	 * Total bytes output from qat de-compression engine.
+	 */
+	kstat_named_t decomp_total_out_bytes;
+	/*
+	 * Number of fails in the qat compression / decompression engine.
+	 * Note: when qat fail happens, it doesn't mean a critical hardware
+	 * issue, sometimes it is because the output buffer is not big enough,
+	 * and the compression job will be transfered to gzip software again,
+	 * so the functionality of ZFS is not impacted.
+	 */
+	kstat_named_t dc_fails;
+
+	/*
+	 * Number of jobs submitted to qat encryption engine.
+	 */
+	kstat_named_t encrypt_requests;
+	/*
+	 * Total bytes sent to qat encryption engine.
+	 */
+	kstat_named_t encrypt_total_in_bytes;
+	/*
+	 * Total bytes output from qat encryption engine.
+	 */
+	kstat_named_t encrypt_total_out_bytes;
+	/*
+	 * Number of jobs submitted to qat decryption engine.
+	 */
+	kstat_named_t decrypt_requests;
+	/*
+	 * Total bytes sent to qat decryption engine.
+	 */
+	kstat_named_t decrypt_total_in_bytes;
+	/*
+	 * Total bytes output from qat decryption engine.
+	 */
+	kstat_named_t decrypt_total_out_bytes;
+	/*
+	 * Number of fails in the qat encryption / decryption engine.
+	 * Note: when qat fail happens, it doesn't mean a critical hardware
+	 * issue, sometimes it is because the output buffer is not big enough,
+	 * and the compression job will be transfered to gzip software again,
+	 * so the functionality of ZFS is not impacted.
```

Comment:
The last sentence needs to be updated to refer to the encyption and not compression job.